### PR TITLE
obinskit: init at 1.1.1

### PR DESCRIPTION
**Pull request by Shou**
###### Motivation for this change

This is the configuration software for the enthusiast Anne Pro mechanical keyboard, used to update the firmware, change keyboard settings, clear Bluetooth connections from the device, set up backlighting colours, and so forth. I'm an owner of one and don't think I'm alone so I thought it'd be convenient to have this packaged. [They have an official Github](https://github.com/obinslab), but the software is proprietary to my knowledge (no license provided).

I've tested that the software works with my own Anne Pro II.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mmahut** [comment]
@GrahamcOfBorg eval

File: pkgs/applications/misc/obinskit/default.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    libpulseaudio
+    libxkbcommon
+    at_spi2_atk
+  ];
+
+in stdenv.mkDerivation rec {
+  pname = "obinskit";
+  version = "1.1.1";
+
+  src = fetchzip {
```
Comment (jonringer) [inline pkgs/applications/misc/obinskit/default.nix]: :)
```suggestion
  src = fetchurl {
```

File: pkgs/applications/misc/obinskit/default.nix [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    url = "http://releases.obins.net/occ/linux/tar/ObinsKit_${version}_x64.tar.gz";
+    sha256 = "0wxsrpv2s917zhx8q5r3p136mj1sbm0adby2vqvi2n6af4bqb9aa";
+  };
+
+  buildInputs = [
+    makeWrapper
+  ];
```
Comment (jonringer) [inline pkgs/applications/misc/obinskit/default.nix]: should be nativeBuildInputs, also you may want to add patchelf

File: pkgs/applications/misc/obinskit/default.nix [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    ln -s $out/opt/obinskit/obinskit $out/bin
+
+    patchelf \
+      --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
+      --set-rpath "$out/lib:${atomEnv.libPath}:${libPath}" \
+      $out/opt/obinskit/obinskit
+
+    wrapProgram $out/opt/obinskit/obinskit \
+      --prefix LD_LIBRARY_PATH : "${atomEnv.libPath}:${libPath}"
+  '';
```
Comment (jonringer) [inline pkgs/applications/misc/obinskit/default.nix]: this is a little redundant with the patchelf command you have, you should be able to remove this

File: pkgs/applications/misc/obinskit/default.nix (hunk 1 above)
Comment (Shou) [reply pkgs/applications/misc/obinskit/default.nix]: I was told `fetchzip` works with archives beyond Zip and indeed it seems to work. Is it still better to use `fetchurl` and then adding an `unpackPhase`?

File: pkgs/applications/misc/obinskit/default.nix (hunk 3 above)
Comment (jonringer) [reply pkgs/applications/misc/obinskit/default.nix]: this is becuase it's done as part of the install phase, do it as part of the postFixup, or set `dontPatchElf = true;` to solve this

Comment (jonringer) [reply pkgs/applications/misc/obinskit/default.nix]: the fixUpPhase likes to mess with rpaths, IIRC

File: pkgs/applications/misc/obinskit/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/applications/misc/obinskit/default.nix]: probably doesn't matter in the end, but you're fetching a tarball, so I would say that a fetchurl is more expected.

File: pkgs/applications/misc/obinskit/default.nix (hunk 3 above)
Comment (Shou) [reply pkgs/applications/misc/obinskit/default.nix]: I've just tried  `patchelf`ing to the `postFixup` phase, and `dontPatchELF = true;`, including both, and no luck. It immediately crashes with:

```
Error: libxkbcommon-x11.so.0: cannot open shared object file: No such file or directory
```

That said, I've remembered that I need to use `strace -f` to get all the processes' outputs, and with that, I've found that I only need to add four libraries to `LD_LIBRARY_PATH` that aren't in `DT_NEEDED`:

```fish
    wrapProgram $out/opt/obinskit/obinskit \
      --prefix LD_LIBRARY_PATH : "${libxkbcommon}/lib:${xlibs.libXt}/lib:${libudev.lib}/lib:${gcc-unwrapped.lib}/lib"
```

And more good news, I've made it run with the `electron_3` Nixpkg binary, which greatly reduces package imports and patching.

Comment (jonringer) [reply pkgs/applications/misc/obinskit/default.nix]: hmm, you're right :(

cc @worldofpeace You're better at wrapping than I am

**Comment by Shou** [comment]
I've removed `libffmpeg.so` and `libnode.so` because `electron_3` already provides these.

File: pkgs/applications/misc/obinskit/default.nix [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    install $src/icudtl.dat $out/opt/obinskit/
+    install $src/natives_blob.bin $out/opt/obinskit/
+    install $src/v8_context_snapshot.bin $out/opt/obinskit/
+    install $src/blink_image_resources_200_percent.pak $out/opt/obinskit/
+    install $src/content_resources_200_percent.pak $out/opt/obinskit/
+    install $src/content_shell.pak $out/opt/obinskit/
+    install $src/ui_resources_200_percent.pak $out/opt/obinskit/
+    install $src/views_resources_200_percent.pak $out/opt/obinskit/
+    cp -r $src/resources $out/opt/obinskit/
+    cp -r $src/locales $out/opt/obinskit/
```
Comment (worldofpeace) [inline pkgs/applications/misc/obinskit/default.nix]: By setting the sourceRoot all these files will be in $CWD so you can drop `$src/` everywhere.

File: pkgs/applications/misc/obinskit/default.nix [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  src = fetchurl {
+    url = "http://releases.obins.net/occ/linux/tar/ObinsKit_${version}_x64.tar.gz";
+    sha256 = "0052m4msslc4k9g3i5vl933cz5q2n1affxhnm433w4apajr8h28h";
+  };
+
+  unpackPhase = ''
+    tar -xzf $src
+    export src='ObinsKit_1.1.1_x64'
+  '';
```
Comment (worldofpeace) [inline pkgs/applications/misc/obinskit/default.nix]: Why is this needed? I seems you want to make sourceRoot be `ObinsKit_1.1.1_x64` so maybe try
```
sourceRoot = "ObinsKit_${version}_x64";
```

and drop this.

File: pkgs/applications/misc/obinskit/default.nix [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    cat <<-EOF > $out/bin/obinskit
+    #!/usr/bin/env bash
+
+    export LD_LIBRARY_PATH="${libxkbcommon}/lib:${xlibs.libXt}/lib:${libudev.lib}/lib:${gcc-unwrapped.lib}/lib"
+
+    exec ${electron_3}/bin/electron $out/opt/obinskit/resources/app.asar
+    EOF
+
+    chmod +x $out/bin/obinskit
```
Comment (worldofpeace) [inline pkgs/applications/misc/obinskit/default.nix]: Using `makeWrapper` will eliminate the need for this entirely.

```
nativeBuildInputs = [
  makeWrapper
];

...

makeWrapper ${electron_3}/bin/electron $out/bin/obinskit \
  --add-flags $out/opt/obinskit/resources/app.asar \
  --prefix LD_LIBRARY_PATH : "${libxkbcommon}/lib:${xlibs.libXt}/lib:${libudev.lib}/lib:${gcc-unwrapped.lib}/lib"
```

also think `gcc-unwrapped` should be `stdenv.cc.cc`. Also `xlibs` is an alias to a new attribute `xorg` so you should use `xorg.libXt`. CI evaluation fails on this pr because of the use of the alias.

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Please set
```
dontConfigure = true;
dontBuild = true;
```

as the build fails trying to run these phases.

Also see my suggestion about using `makeWrapper` and the use of aliases in this pr.

**Comment by Shou** [comment]
@worldofpeace thanks for the CR, I should've taken care of the feedback now, and squashed.

**Review by jonringer — APPROVED** [review]
diff LGTM
commit LGTM
app opens fine

```
[4 built, 2 copied (123.9 MiB), 36.1 MiB DL]
https://github.com/NixOS/nixpkgs/pull/71259
1 package built:
obinskit
```
